src/app.ts

```diff
@@ -85,8 +88,38 @@ client.on('interactionCreate', async interaction => {
         server = await JoinGuild(interaction.guild)
     }
     await server.EnsureHasRole(interaction.member as GuildMember)
+    if(interaction.isCommand()) {
+        await ProcessCommand(server, interaction)
+    }
+    if(interaction.isButton()) {
+        //a space separates the type of interaction and the name of the queue channel
+        const type = interaction.customId.split(" ")[0]
+        const queue_name = interaction.customId.split(" ")[1]
+
+        if(!(interaction.member instanceof GuildMember)) {
+            await interaction.reply({content: `Erm. Somethings wrong, this shouldn't happen. I'll inform the humaniod that maintains me`, ephemeral: true})
+            console.error(`Recieved an interaction without a member from user ${interaction.user} on server ${interaction.guild}`)
+            return;
+        }
 
-    await ProcessCommand(server, interaction)
+        if(type === 'join') {
+            await server.EnqueueUser(queue_name, interaction.member, interaction.type).catch((errstr : Error) => {
+                if(interaction.member instanceof GuildMember) {
+                    interaction.member.send(errstr.message)
+                }
+            })
+        } else if(type === 'leave') {
+             await server.RemoveMember(queue_name, interaction.member, interaction.type).catch((errstr : Error) => {
+                if(interaction.member instanceof GuildMember && errstr.name == 'UserError') {
+                    interaction.member.send(errstr.message)
+                }
+             })
+        } else {
+            console.error('Received invalid button interaction')
+        }
+        //interaction.deferUpdate()
```

Comment:
nit: Remove this comment

Reply:
this comment was left on purpose. it currently causes the bot to crash if it takes too long to do a enqueue or leave when I run the bot locally (my laptop is pretty slow when running bots). This line of code should remove the "this interaction failed" error that shows up after you press the button.

Reply:
Deferring the reply makes sense, that's what happens in the `ProcessCommand` path. It would make sense to defer the button interaction update as well. Lets uncomment this and put it as the first line after `if(type == 'join')`